### PR TITLE
Refactor extension organization

### PR DESCRIPTION
**Pull request by kgryte**
PR to address https://github.com/jupyterlab/jupyterlab-data-explorer/issues/83.

### CODE REVIEW

File: dataregistry-extension/src/csvviewer/activate.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
+/**
```
Comment (saulshanabrook) [inline dataregistry-extension/src/csvviewer/activate.ts]: Instead of `activate.ts` maybe have `converters.ts` file that exports function from `registry` to array of converters.

File: dataregistry-extension/src/csvviewer/data_grid.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
+/**
```
Comment (saulshanabrook) [inline dataregistry-extension/src/csvviewer/data_grid.ts]: Maybe call it CSVDataGrid to match the class? What do you think?

File: dataregistry-extension/src/csvviewer/data_grid.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    this._subscription = this._data.subscribe(onData);
+    super.onBeforeAttach(msg);
+
+    /**
+     * Callback invoked upon changes to the CSV data.
+     *
+     * @private
+     * @param data - CSV data
+     */
+    function onData(data: string) {
```
Comment (saulshanabrook) [inline dataregistry-extension/src/csvviewer/data_grid.ts]: Maybe make this a method?

File: dataregistry-extension/src/csvviewer/data_grid.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    this._subscription && this._subscription.unsubscribe();
+    super.onBeforeDetach(msg);
+  }
+}
+
+/**
+ * Exports.
+ */
+export { CSVDataGrid };
+export default CSVDataGrid;
```
Comment (saulshanabrook) [inline dataregistry-extension/src/csvviewer/data_grid.ts]: I don't think. we need a default export

File: dataregistry-extension/src/mimetypes.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/**
```
Comment (saulshanabrook) [inline dataregistry-extension/src/mimetypes.ts]: I don't think we need a mimetypes file

File: dataregistry-extension/src/datatypes.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  "notebookContext": new DataTypeNoArgs<Observable<Context<INotebookModel>>>(mimetypes.notebookContext),
+  "text": new DataTypeStringArg<Observable<string>>("text/plain", "mimeType"),
+  "textContext": new DataTypeStringArg<Observable<Context<DocumentRegistry.ICodeModel>>>(mimetypes.textContext, "mimeType"),
+  "widget": new DataTypeStringArg<() => Widget>(mimetypes.widget, "label")
+};
+
+/**
+ * Exports.
+ */
+export { datatypes };
```
Comment (saulshanabrook) [inline dataregistry-extension/src/datatypes.ts]: I think this should just export each data type, like:

```
export const CSVDataType = new DataTypeNoArgs<Observable<string>>("text/csv");
```

We don't need to copy all these other ones. They are defined in other files.

File: dataregistry-extension/src/csvviewer/index.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ * @license BSD-3-Clause
+ *
+ * Copyright (c) 2019 Project Jupyter Contributors.
+ * Distributed under the terms of the 3-Clause BSD License.
+ */
+
+import { JupyterFrontEndPlugin } from "@jupyterlab/application";
+import { IRegistry } from "./../registry";
+import { activate } from "./activate";
+
```
Comment (saulshanabrook) [inline dataregistry-extension/src/csvviewer/index.ts]: I think this should also export all from the datatypes file, so that other modules can refer to this datatypes

File: dataregistry-extension/src/csvviewer/activate.ts (hunk 1 above)
Comment (kgryte) [reply dataregistry-extension/src/csvviewer/activate.ts]: My motivation: (1) 1 file, 1 concern and (2) testing.
​
In short, a guiding principle in Node.js development is to create files (modules) which export one thing (i.e., have a single concern), in order to encourage readability and testability.
​
If a file contains a single export, no need to track down a file's exports, which, if a file is long and not structured, can be tedious and time consuming.
​
And if a file has a single export, encourages writing tests which import that single export and directly test an API surface area. Often when everything is in long files, we see correspondingly long test files with correspondingly increased difficulty in achieving 100% code coverage. By having single file tests, easier to test using mocks via dependency injection (classic is proxyquire to proxy imported dependencies) and, as tests are documentation, easier to see how exported functionality should behave.
​
For example, take `activate.ts`. The file exports a single function which accepts two arguments and whose intent is to register "converters". When testing, we can import the sole export from this individual file, and, as Jupyter uses dependency injection, we can mock the second argument with an object containing a single method: `addConverter`, testing whether particular converters are registered, etc.
​
Similarly, I moved the converters to separate files based on the same idea. We should be able to test a particular converter, say, the one generated by export from `text_to_csv`, to ensure proper conversion behavior, such as the conversion of text "observable" to a CSV "observable", etc. If everything is in a single file and all the logic tightly coupled, this becomes rather difficult to test, especially thoroughly.
​
In a sense, each file having a single concern allows for, perhaps arguably, easier enforcement of API contracts.
​
Single-file-single-concern "style" can be found, to some degree, in Node.js core, and more so in Node.js modules written with the file system in mind and not, say, like jQuery, as a single file to be downloaded by a browser.
​
This "style" takes inspiration from the Unix philosophy in the sense that each module should do only one thing and do that thing well. Often, we think of this at the package level, but applies equally to the module/file layer.

File: dataregistry-extension/src/csvviewer/data_grid.ts (hunk 2 above)
Comment (kgryte) [reply dataregistry-extension/src/csvviewer/data_grid.ts]: Can do, if think it is useful. The main idea here is that, to a module consumer, an implementation (and name) should be opaque and having to know an exported symbol identifier means having to know something about an implementation (namely, the identifiers in that file).
​
Here, my desire in the `csv_to_data_grid.ts` file is to convert from CSV to a data grid. The activation file does not need to know that the `data_grid.ts` file exports an extended class (i.e., a specialized data grid). To it, it can simply import from `data_grid.ts` a class for creating a data grid widget. So the naming of just `data_grid.ts` was to reflect that idea: from the `data_grid.ts` file, I can import a class for creating a data grid, assuming that whatever that class is can handle CSV data.
​
I updated the exported name in `data_grid.ts` to reflect this idea.

File: dataregistry-extension/src/csvviewer/data_grid.ts (hunk 3 above)
Comment (kgryte) [reply dataregistry-extension/src/csvviewer/data_grid.ts]: Yes, this could make sense. Generally, I avoid hanging private (implementation) methods off prototypes to ensure encapsulation, unless their are clear performance advantages to do otherwise.
​
Which raises the question: how frequently is the `before-attach` message sent? If often, then I should definitely create an private internal method; but, if not, then probably fine to leave as is.

File: dataregistry-extension/src/csvviewer/data_grid.ts (hunk 4 above)
Comment (kgryte) [reply dataregistry-extension/src/csvviewer/data_grid.ts]: In general, my preference would be that every file have **only** a default export and no named exports for reasons mentioned earlier. However, I mainly included named exports because, with few exceptions, named exports are the norm in this project.
​
This has been debated somewhat in the Node.js community where some developers do both in order to allow users to pick whichever style they prefer and whichever style matches their project conventions.

File: dataregistry-extension/src/mimetypes.ts (hunk 5 above)
Comment (kgryte) [reply dataregistry-extension/src/mimetypes.ts]: The reason I did was because I noticed MIME type magic constants which were repeated in various places (e.g., `if type === "text/csv" then...`).
​
Whenever I see repeated magic constants, I sense we'd be better off defining these in a single place (file) which we can then import.
​
Otherwise, we run the risk of typos and subtle bugs. For example, if I forget the `i` in `"text/plain"` when doing a comparison or when providing as an argument, TypeScript won't be able to catch this as a type error. I'd still be comparing strings or providing a string argument.
​
In this case, I'd only be able to catch the typo during runtime tests, not at compile time.
​
However, if I import a MIME types database and misspell a property name (most likely a name which does not exist), then I am more likely to get a compile error because I'd provide, e.g., `undefined` as an argument and not a string, and avoid the runtime bug.

File: dataregistry-extension/src/datatypes.ts (hunk 6 above)
Comment (kgryte) [reply dataregistry-extension/src/datatypes.ts]: There may be some confusion here. This file is not defined in the `csvviewer` folder by in the parent directory.
​
In short, the idea is that, rather than having each data converter extension export various data types, thus requiring us to read individual files to see what data types are available, we should provide a single "database" of datatypes.
​
This allows us, when creating converters, to only need to consult one file, rather than many, in order to know what datatypes are implemented in the dataregistry.
​
If we have datatypes scattered across different files, we'll have to engage in an expedition to find files exporting data types and then, even supposing we codify a file in which these are defined, we have to read that separate files to know the identifier name which is exported (e.g., `CSVDataType` or `DatasetsFileDataType`, etc).
​
Instead, I'd argue that we'd be better off centralizing the declaration of data types in a single file (i.e., database). Meaning, in this refactoring, I will move the data types declarations to this `datatypes.ts` file. Then, whenever I'd want to implement a new converter, I'd only need to consult a single file.

**Comment by kgryte** [comment]
@saulshanabrook This PR still needs some small changes, but the broad strokes are there. Obviously, many files have changed. I wrote up a development guide and setup a number of `npm` scripts.

I think it would be beneficial to get your feedback, if any. You may want to pull this down separately, follow the development guide, and see if you can run everything locally.

My sense is that, once I add a few touch-ups, would be good to merge this as is, so that we can begin working from the reorganized structure. I can move the remaining converters to individual extension packages in a piecemeal fashion.

**Comment by saulshanabrook** [comment]
>  I  think it would be beneficial to get your feedback, if any. You may want  to pull this down separately, follow the development guide, and see if  you can run everything locally. My sense is that, once I add a few touch-ups, would be good to merge  this as is, so that we can begin working from the reorganized structure.  I can move the remaining converters to individual extension packages in  a piecemeal fashion.


Sounds good, I will give this a go.

File: package.json [hunk 8]
```diff
@@ -1,16 +1,46 @@
...
+    "lint": "jlpm run lint:css && jlpm run lint:typescript",
+    "lint:css": "stylelint packages/**/*.css",
+    "lint:typescript": "eslint packages/**/src/*.ts",
+    "rebuild:packages": "jlpm run clean:packages && jlpm run build:packages",
+    "registry": "verdaccio",
+    "registry:init": "npm adduser --registry http://localhost:4873 && lerna exec --scope dataregistry-csvviewer --scope dataregistry-registry -- npm publish --registry http://localhost:4873",
+    "uninstall:extensions": "jupyter labextension uninstall --all --no-build",
+    "unlink:packages": "lerna run unlink"
+  },
+  "devDependencies": {
```
Comment (saulshanabrook) [inline package.json:34]: Do you have a sense of what dev dependencies should be in the root package vs the child packages? It seems like some are added in both places.

**Comment by saulshanabrook** [comment]
So far I have been using prettier on this codebase to maintain a consistent style. Can I go ahead and reformat all of your work with it?

File: package.json (hunk 8 above)
Comment (kgryte) [reply package.json:34]: Yeah, I can remove `rimraf` and `typescript` from the individual packages, I believe.

**Comment by kgryte** [comment]
I did not see a build command for reformatting, so I had yet to do so.

File: docs/development.md [hunk 9]
```diff
@@ -0,0 +1,264 @@
+<!--
```
Comment (saulshanabrook) [inline docs/development.md:1]: Thoughts on moving this to `CONTRIBUTING.md` so [github picks it up](https://help.github.com/en/articles/setting-guidelines-for-repository-contributors#adding-a-contributing-file)?

**Comment by saulshanabrook** [comment]
> I did not see a build command for reformatting, so I had yet to do so.

Yeah there isn't one. We should add one, at least for linting to verify everything is prettified. I have my editor set up to reformat on save.

File: docs/development.md (hunk 9 above)
Comment (kgryte) [reply docs/development.md:1]: I typically view the `CONTRIBUTING.md` file to be a bit higher level, linking to code of conducts, including a developer's certificate of origin, and linking to other docs necessary for contributing, such as a developer environment guide, style guides, etc. For an example, see [here](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md). On reflection, some overlap between them. I am fine with consolidating if you think it best.

**Comment by kgryte** [comment]
I'd say that, ideally, we'd make such development processes editor agnostic. :)

**Comment by saulshanabrook** [comment]
> I'd say that, ideally, we'd make such development processes editor agnostic. :)

Yep that sounds ideal :)

File: docs/development.md (hunk 9 above)
Comment (saulshanabrook) [reply docs/development.md:1]: I am fine with merging this as is and iterating. I am fine with either.

File: docs/development.md [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+-   [Python][python]: general purpose language (version `>= 3.6`)
+-   [pip][pip]: Python package manager (version `>= 9.0.0`)
+-   [Node.js][node-js]: JavaScript runtime (latest stable version is **strongly** recommended)
+-   [npm][npm]: package manager (version `> 2.7.0`)
+-   [JupyterLab][jupyterlab]: computational environment (version `>= 1.0.0`)
+
+While not required, you are encouraged to create an [Anaconda][anaconda] environment.
+
+```bash
+$ conda create -n jupyterlab-data-explorer -c conda-forge python=3.6 -c anaconda jupyterlab nodejs
```
Comment (saulshanabrook) [inline docs/development.md]:
```suggestion
$ conda create -n jupyterlab-data-explorer -c conda-forge python=3.6 jupyterlab nodejs
```

File: docs/development.md (hunk 9 above)
Comment (kgryte) [reply docs/development.md:1]: Added a contributing doc. In the abstract, I see "contributing" and "development" as two separate, but possible overlapping things. For example, I may want to locally develop, but never contribute, and I may want to contribute, but never develop. The contributing guide I included is based on the one I linked to above and includes a bit more hand holding for neophytes. Maybe this is unnecessary.

Comment (saulshanabrook) [reply docs/development.md:1]: That makes sense to me! Thank you.

File: docs/development.md (hunk 10 above)
Comment (kgryte) [reply docs/development.md]: Okay. Question. I essentially took this from the Jupyter [docs](https://jupyterlab.readthedocs.io/en/stable/developer/extension_tutorial.html#install-nodejs-jupyterlab-etc-in-a-conda-environment). The `-c anaconda` is specifying that the remaining arguments be installed via the  Anaconda "channel", correct? If this is unnecessary, why do the Jupyter docs show otherwise (or maybe I am misreading them)?

Comment (saulshanabrook) [reply docs/development.md]: I believe you have to put all the channel names first. In the jupyter docs, it does this:

```
conda create -n jupyterlab-ext --override-channels --strict-channel-priority -c conda-forge -c anaconda jupyterlab cookiecutter nodejs git
```

I always prefer to just to use conda-forge and not the anaconda channel. I know we had a long discussion about this over slack at quansight just this week... @scopatz Do you have any recommendations on including anaconda channel or just conda forge in this command?

Comment (kgryte) [reply docs/development.md]: Updated.

**Comment by saulshanabrook** [comment]
What do you think about having the packages that are extensions end in `-extension`? That's been our custom in JupyterLab core. So `dataregistry-csvviewer` -> `dataregistry-csvviewer-extension` and `dataregistry-registry` -> `dataregistry-registry-extension`?

File: package.json [hunk 11]
```diff
@@ -1,16 +1,46 @@
...
+    "clean:jupyter": "jlpm run uninstall:extensions && (lerna run --no-bail unlink || echo 'At least one unlink command failed, but continuing...')",
+    "clean:node": "lerna clean --yes && rimraf node_modules",
+    "clean:packages": "lerna run clean",
+    "link:packages": "lerna run link",
+    "lint": "jlpm run lint:css && jlpm run lint:typescript",
+    "lint:css": "stylelint packages/**/*.css",
+    "lint:typescript": "eslint packages/**/src/*.ts",
+    "rebuild:packages": "jlpm run clean:packages && jlpm run build:packages",
+    "registry": "verdaccio",
+    "registry:init": "npm adduser --registry http://localhost:4873 && lerna exec --scope dataregistry-csvviewer --scope dataregistry-registry -- npm publish --registry http://localhost:4873",
```
Comment (saulshanabrook) [inline package.json]: I had to change this line to specify the package names instead of the paths:

```suggestion
    "registry:init": "npm adduser --registry http://localhost:4873 && lerna exec --scope @jupyterlab/dataregistry-csvviewer --scope @jupyterlab/dataregistry-registry -- npm publish --registry http://localhost:4873",
```

Otherwise I would get:

```
jlpm run registry:init
yarn run v1.15.2
$ npm adduser --registry http://localhost:4873 && lerna exec --scope dataregistry-csvviewer --scope dataregistry-registry -- npm publish --registry http://localhost:4873
Username: abc
Password:
Email: (this IS public) abc@abc.com
Logged in as abc on http://localhost:4873/.
lerna notice cli v3.16.4
lerna info versioning independent
lerna info filter [ 'dataregistry-csvviewer', 'dataregistry-registry' ]
lerna ERR! EFILTER No packages remain after filtering [ 'dataregistry-csvviewer', 'dataregistry-registry' ]
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by kgryte** [comment]
I thought about that, but then disliked how long the package names were getting. However, I recognize that this seems to be a growing convention within Jupyter and the ecosystem. I'll go with whatever you think best on this, as you have a better pulse on where things are going and the attendant conventions.

File: package.json (hunk 11 above)
Comment (kgryte) [reply package.json]: Ah, you're right! I misunderstood the scope filters. Will fix now.

Comment (kgryte) [reply package.json]: Pushed a patch.

**Comment by saulshanabrook** [comment]
> I thought about that, but then disliked how long the package names were getting.

I think it's useful, just as another signal to developers which things expose an extension and which do not.

**Comment by kgryte** [comment]
Rog. Will rename shortly.

File: packages/dataregistry-csvviewer/package.json [hunk 12]
```diff
@@ -37,6 +37,7 @@
     "@jupyterlab/application": "^1.0.0",
     "@jupyterlab/csvviewer": "^1.0.0",
     "@jupyterlab/dataregistry": "^3.0.0",
+    "@jupyterlab/dataregistry-registry": "^0.0.0",
```
Comment (kgryte) [inline packages/dataregistry-csvviewer/package.json]: See L44. This is already included.

Comment (saulshanabrook) [reply packages/dataregistry-csvviewer/package.json]: Thanks!

File: package.json [hunk 13]
```diff
@@ -1,16 +1,48 @@
...
-    "build:dev": "jlpm install && jlpm clean && jlpm run build && jlpm run link",
-    "lint":  "jlpm workspace @jupyterlab/dataregistry-extension lint",
-    "clean": "jlpm workspace @jupyterlab/dataregistry run clean && jlpm workspace @jupyterlab/dataregistry-extension run clean",
-    "build:watch": "tsc --build --watch  dataregistry dataregistry-extension",
-    "link": "jupyter labextension link dataregistry --no-build && jupyter labextension link dataregistry-extension",
-    "unlink": "jupyter labextension unlink dataregistry --no-build && jupyter labextension unlink dataregistry-extension"
+    "all": "jlpm run clean && jlpm run build && jlpm run build:jupyter",
+    "all:watch": "jlpm run clean && jlpm run build && jlpm run build:jupyter:watch",
+    "build": "jlpm run build:dev",
+    "build:dev": "jlpm install && jlpm run clean:packages && lerna link && jlpm run build:packages && jlpm run link:packages",
```
Comment (saulshanabrook) [inline package.json]: Any reason  to use `lerna link` over `lerna bootstrap`? I don't know the difference but the lerna readme just mentions bootstrap.

Comment (kgryte) [reply package.json]: Yeah, did not want Lerna executing any of the other commands in `bootstrap` beside linking.

Comment (kgryte) [reply package.json]: See [here](https://github.com/lerna/lerna/tree/master/commands/bootstrap#usage).

**Comment by saulshanabrook** [comment]
Oh no, I am getting the dreaded error!

> error Couldn't find package "@jupyterlab/dataregistry-registry@0.0.0" required by "@jupyterlab/dataregistry-csvviewer@file:../extensions/jupyterlab-dataregistry-csvviewer-0.0.0-718c020eb361908eddb162f4fc6b1ccd253c90ba.tgz" on the "npm" registry.

I think the registry isn't being passed in through the env variable for me, since it isn't hitting the API

File: package.json (hunk 13 above)
Comment (saulshanabrook) [reply package.json]: Ah thank you! Sounds good

**Comment by kgryte** [comment]
And now I am encountering the same error. **sigh**

File: package.json [hunk 14]
```diff
@@ -1,16 +1,50 @@
...
-    "build:watch": "tsc --build --watch  dataregistry dataregistry-extension",
-    "link": "jupyter labextension link dataregistry --no-build && jupyter labextension link dataregistry-extension",
-    "unlink": "jupyter labextension unlink dataregistry --no-build && jupyter labextension unlink dataregistry-extension"
-  }
+    "all": "jlpm run clean && jlpm run build && jlpm run build:jupyter",
+    "all:watch": "jlpm run clean && jlpm run build && jlpm run build:jupyter:watch",
+    "build": "jlpm run build:dev",
+    "build:dev": "jlpm install && jlpm run clean:packages && lerna link && jlpm run build:packages && jlpm run link:packages && jupyter labextension list",
+    "build:jupyter": "jupyter lab build || (jlpm run build:jupyter:remediate && jupyter lab build)",
+    "build:jupyter:remediate": "curr_dir=$PWD && (cd $(jupyter lab path | head -n1 | sed s/Application\\ directory:\\ *//)/staging && echo $PWD && echo '\"@jupyterlab:registry\" \"http://localhost:4873\"' >> .yarnrc && cat .yarnrc && YARN_REGISTRY=http://localhost:4873 node ./yarn.js install --non-interactive && cd $curr_dir) || cd $curr_dir",
```
Comment (saulshanabrook) [inline package.json:14]: Nothing will go wrong here! :)

File: docs/development.md (hunk 10 above)
Comment (scopatz) [reply docs/development.md]: yeah, I recommend conda-forge

File: package.json (hunk 14 above)
Comment (kgryte) [reply package.json:14]: Nope. Nothing could go wrong. It would be better if we only performed the "remediation" step when we encounter the npm registry error, but that would require parsing the error log, which seems a bit heavy handed atm. Assuming we encounter a non-registry error which is reproducible, the second `jupyter lab build` should simply generate the same error. Or at least that is the hope...

File: packages/dataregistry/src/cachedObservable.ts [hunk 15]
```diff
@@ -139,7 +139,7 @@ export class CachedObservable<T> extends Observable<T> {
           subscriber.next(state.value);
           return () => {};
         default:
-          const _exhaustiveCheck: never = state;
```
Comment (saulshanabrook) [inline packages/dataregistry/src/cachedObservable.ts:142]: Nice! This is great, since it also preserves verifying at compile time that  the check is exhaustive. Thanks.

**Comment by saulshanabrook** [comment]
Unfortunately, I am still getting the error. I think the issue is that after you add the yarnrc we run jupyter lab build, which wipes it out.

**Comment by saulshanabrook** [comment]
This has been coming up for others as well, I am going to bring it up at tomorrows dev meeting: https://github.com/jupyterlab/jupyterlab#weekly-dev-meeting

File: packages/dataregistry-csvviewer-extension/src/datatypes.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * Copyright (c) 2019 Project Jupyter Contributors.
+ * Distributed under the terms of the 3-Clause BSD License.
+ */
+
+import { DataTypeNoArgs, DataTypeStringArg } from "@jupyterlab/dataregistry";
+import { Widget } from "@phosphor/widgets";
+import { Observable } from "rxjs";
+
+// TODO: refactor
```
Comment (saulshanabrook) [inline packages/dataregistry-csvviewer-extension/src/datatypes.ts:12]: This looks good to me at the moment. The datatypes will be properly typed, so if you do `datatypes.csv` it will know the type. What do you have in mind for refactor?

File: packages/dataregistry-csvviewer-extension/src/csv_to_data_grid.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+ * @private
+ * @returns data type converter
+ */
+function csv2datagrid(): any {
+  const conversion = {
+    "from": datatypes.csv,
+    "to": datatypes.widget
+  };
+
+  return createConverter(conversion, convert);
```
Comment (saulshanabrook) [inline packages/dataregistry-csvviewer-extension/src/csv_to_data_grid.ts:61]: I prefer the functions inline here, since the typescript know the types of everything. Otherwise, you have to manually specify the types. Like below, `convert` has no return annotation, so it is not verified that it is correct. It's a pain to add those annotations so if we have the functions inline, then we dont have to.

File: packages/dataregistry-csvviewer-extension/src/text_to_csv.ts [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+import { createConverter } from "@jupyterlab/dataregistry";
+import datatypes from "./datatypes"; // FIXME
+
+/**
+ * Interface describing an object containing data to convert.
+ *
+ * @private
+ */
+interface Data {
```
Comment (saulshanabrook) [inline packages/dataregistry-csvviewer-extension/src/text_to_csv.ts:16]: Again here, since the functions are not inline you have to manually annotate them with types, and this type is too general.

**Comment by saulshanabrook** [comment]
At long last, the build succeeds! I followed the dev instructions and it worked. I also tested changing some code using the watch commands and that worked as well. I just left a few more comments, about some type annotations.

TLDR: Keeping callbacks inline means that you don't have to define type annotations for them, which keeps code concise and typed strongly.

File: packages/dataregistry-csvviewer-extension/src/datatypes.ts (hunk 16 above)
Comment (kgryte) [reply packages/dataregistry-csvviewer-extension/src/datatypes.ts:12]: I have the TODO here as this is a temporary hack until each of the converters is a separate extension. E.g., instead of L26, we'd import the data type from the widget data converter extension.

File: packages/dataregistry-csvviewer-extension/src/csv_to_data_grid.ts (hunk 17 above)
Comment (kgryte) [reply packages/dataregistry-csvviewer-extension/src/csv_to_data_grid.ts:61]: Yeah, I ended up frustrated with inlining because you cannot readily document inlined functions and cannot elevate them to a higher scope.

I also look at this another way:

> what is the minimum contract necessary for the `convert` function to work and for someone (e.g., an outside developer reading the source code for this extension) to be able to grok ?

Based on how `createConverter` works, I can already assume that `convert` is going to be provided an object containing data of the correct "type", that is part of the `createConverter` "contract", and we've already defined in the `conversion` object the data type in the return object. 

In short, TMK, we don't lose any "correctness" by using general types for what is an internal function interface.

And personally, and this is one of my gripes about using TypeScript, I shouldn't have to choose between precise and exhaustive types and no documentation/scope hoisting at all. TypeScript effectively encourages devs to inline functions (often via arrow functions in order to avoid explicit typing) which is not, IMO, good practice for a variety of reasons (perf, docs, debugging, readability, etc).

This stated, I can modify to the conventions of this project, if this is a blocker for merge. :)

Comment (saulshanabrook) [reply packages/dataregistry-csvviewer-extension/src/csv_to_data_grid.ts:61]: Thanks for the explanation here, that makes sense! I haven't thought as much about the downsides of Typescript's strictness, I just let it tell me what to do :)

Let's merge this in.
